updating variables (#83)
infinum/eightshift-docs: website/docs/basics/blocks-styles.md

---
id: blocks-styles
title: Styles
sidebar_label: Styles
---

[![docs-source](https://img.shields.io/badge/source-eigthshift--frontend--libs-yellow?style=for-the-badge&logo=javascript&labelColor=2a2a2a)](https://github.com/infinum/eightshift-frontend-libs/tree/develop/blocks/init/src/blocks/)

> We decided to remove all support for IE 11 and at that point, a bunch of doors opened up for new features and cool stuff. One of those features is CSS variables with block manifest and global settings.

Reviewer: ```suggestion
We decided to remove all support for IE 11.  That decision opened up doors that lead to new, awesome features and all the cool stuff. One of those features is CSS variables with block manifest and global settings.
```

Let's dig into the implementation of CSS variables in your project.

## Setup

To be able to use CSS variables you need to implement two helpers in your blocks/components: `outputCssVariables` and `getUnique`. For more details on these helpers, you can read [here](helpers-javascript#outputcssvariables). 

> There are a few differences between the JavaScript and PHP implementation due to how React handles component re-rendering.

Reviewer: ```suggestion
There are a few differences between the JavaScript and PHP implementation due to a way React handles component re-rendering.
```

**typography-editor.js**
```js
import React, { useMemo } from 'react';
import { outputCssVariables, getUnique } from '@eightshift/frontend-libs/scripts/editor';
import manifest from '../manifest.json';

export const TypographyEditor = (attributes) => {

  // We need to use memo in order to optimise component re-rendering.
  const unique = useMemo(() => getUnique(), []);

  return (
    <>
      {outputCssVariables(attributes, manifest, unique)}

      <div data-id={unique}>
        {/*Regular component implementation*/}
      </div>
    </>
  );
};
```

**typography.php**
```php
/**
 * Template for the Typography Component.
 *
 * @package Eightshift_Boilerplate
 */

use EightshiftLibs\Helpers\Components;

$manifest = Components::getManifest(__DIR__);

$unique = Components::getUnique();
echo Components::outputCssVariables($attributes, $manifest, $unique); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped

?>

<div data-id="<?php echo esc_attr($unique); ?>">
  // Regular component implementation
</div>
```

> Now that we have helpers in place let's see how they work and what features they offer.

Reviewer: ```suggestion
Now that we have helpers in place, let's see how they work and what features they offer.
```

## Details

> CSS variables helper consists of 2 helpers. `outputCssVariables` helper will output all CSS variables set in your blocks/component manifest and `getUnique` helper will make sure that variables are applied only to the correct block/component.

Reviewer: ```suggestion
CSS variables helper consists of 2 helpers. `outputCssVariables` helper will output all CSS variables set in your blocks/component manifest and `getUnique` helper will make sure that variables are applied only to the exact block/component.
```

**Manifest:**
```json
{
  "attributes": {
    "typographySize": {
      "type": "string",
      "default": "120-default",
      "variable": "default"
    }
  },
  "options": {
    "typographySize": [
      {
        "label": "180 Display",
        "value": "180-default"
      },
      {
        "label": "120 Display",
        "value": "120-default"
      },
      {
        "label": "80 Display",
        "value": "80-default"
      }
    ]
  }
}
```

**Output:**
```css
--typography-size: 120-default
```

## Color variables

If you have an attribute that represent colors from global variables, you can output the CSS variable for color directly.

**Manifest:**
```json
{
  "attributes": {
    "typographyColor": {
      "type": "string",
      "default": "white",
      "variable": "color"
    }
  },
  "options": {
    "typographyColor": [
      {
        "name": "White",
        "slug": "white",
        "color": "#ffffff"
      },
      {
        "name": "Black",
        "slug": "black",
        "color": "#111111"
      },
    ]
  }
}
```

**Output:**
```css
--typography-color: var(--global-colors-white);
```

> ## Complex attributes

Reviewer 2: ```suggestion
## Complex attribute variables
```

If you have attributes stored as objects (for example used in the `SelectControl` component), by default CSS variable generator will output the `value` stored in the attribute.
Setting the `variable` key to `select`, the CSS variable generator will look for the `variable` key in the manifest `options` field.

If the `variable` key is not set in options, the `value` key will be used as fallback.

> You can write any CSS declaration or variable in the custom variable key.

**Manifest:**
```json
{
  "attributes": {
    "typographySize": {
      "type": "string",
      "default": "120-default",
      "variable": "select"
    }
  },
  "options": {
    "typographySize": [
      {
        "label": "180 Display",
        "value": "180-default",
        "variable": "something"
      },
      {
        "label": "120 Display",
        "value": "120-default",
        "variable": "somethingNew"
      },
      {
        "label": "80 Display",
        "value": "80-default",
        "variable": "somethingNew2"
      }
    ]
  }
}
```

**Output:**
```css
--typography-size: somethingNew
```

> ### Complex variables - Editor only

Reviewer 2: ```suggestion
### Complex variables inside the Block editor
```

---
id: blocks-styles
title: Styles
sidebar_label: Styles
---

[![docs-source](https://img.shields.io/badge/source-eigthshift--frontend--libs-yellow?style=for-the-badge&logo=javascript&labelColor=2a2a2a)](https://github.com/infinum/eightshift-frontend-libs/tree/develop/blocks/init/src/blocks/)

We decided to remove all support for IE 11 and at that point, a bunch of doors opened up for new features and cool stuff. One of those features is CSS variables with block manifest and global settings.

Let's dig into the implementation of CSS variables in your project.

## Setup

> To be able to use CSS variable you need to implement two helpers `outputCssVariables` and `getUnique` in your block/component. For more details on these helpers, you can read [here](helpers-javascript#outputcssvariables). 

Reviewer 2: ```suggestion
To be able to use CSS variables you need to implement two helpers in your blocks/components: `outputCssVariables` and `getUnique`. For more details on these helpers, you can read [here](helpers-javascript#outputcssvariables). 
```

> There are a few small differences between implementation in JavaScript and PHP due to how React handles components re-rendering.

Reviewer 2: ```suggestion
There are a few differences between the JavaScript and PHP implementation due to how React handles component re-rendering.
```

**typography-editor.js**
```js
import React, { useMemo } from 'react';
import { outputCssVariables, getUnique } from '@eightshift/frontend-libs/scripts/editor';
import manifest from '../manifest.json';

export const TypographyEditor = (attributes) => {

>   // We need to use memo in order to optimise components re-render.

Reviewer 2: ```suggestion
  // We need to use memo in order to optimise component re-rendering.
```

  const unique = useMemo(() => getUnique(), []);

  return (
    <>
      {outputCssVariables(attributes, manifest, unique)}

      <div data-id={unique}>
        {/*Regular component implementation*/}
      </div>
    </>
  );
};
```

**typography.php**
```php
/**
 * Template for the Typography Component.
 *
 * @package Eightshift_Boilerplate
 */

use EightshiftLibs\Helpers\Components;

$manifest = Components::getManifest(__DIR__);

$unique = Components::getUnique();
echo Components::outputCssVariables($attributes, $manifest, $unique); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped

?>

<div data-id="<?php echo esc_attr($unique); ?>">
  // Regular component implementation
</div>
```

> Now that we have helpers in place let's see how they are working and what features they have to offer.

Reviewer 2: ```suggestion
Now that we have helpers in place let's see how they work and what features they offer.
```

## Details

> CSS variables helper work in a combination of 2 helpers. `outputCssVariables` helper will output all CSS variables set in your blocks/component manifest and `getUnique` helper will make sure that variables are only applied to the correct block/component.

Reviewer 2: ```suggestion
CSS variables helper consists of 2 helpers. `outputCssVariables` helper will output all CSS variables set in your blocks/component manifest and `getUnique` helper will make sure that variables are applied only to the correct block/component.
```

If you check your rendered website you can see that you have something like this:
```html
<style>
  .typography[data-id='210c9bbf733ef5c6aa74c49168ac29a7'] {
    --typography-color: var(--global-colors-black);
    --typography-content-align: left;
  }
</style>

<div class="typography" data-id="210c9bbf733ef5c6aa74c49168ac29a7">...</div>
```

> This is it for all the magic of CSS variables. Now let us see it in action and all those additional features you can utilize.

Reviewer 2: ```suggestion
This is all that's required for the magic of CSS variables to work! Now let's see it in action and check out all additional features you can use.
```

## Global variables

> Now that we know how CSS variables are outputted we have one additional helper to describe. This helper is loaded in `application-blocks-editor.js` file and it is called `outputCssVariablesGlobal(globalSettings);`.

Reviewer 2: ```suggestion
Now that we know how CSS variables are generated, we have one more helper to describe.
```

> It is used to output all CSS variables from the global manifest under the `globalVariables` key to `:root` selector.

Reviewer 2: ```suggestion
The `outputCssVariablesGlobal(globalSettings);` helper is called in the `application-blocks-editor.js` file. It is used to output all CSS variables from the global manifest under the `globalVariables` key, into the `:root` selector.
```

> > Those variables are available in all your blocks/components.

Reviewer 2: ```suggestion
> All of these variables are available to use inside your blocks/components.
```

**Global Manifest:**
```json
{
  "namespace": "eightshift-boilerplate",
  "background": "#D8262C",
  "foreground": "#FFFFFF",
  "globalVariables": {
    "customBlocksName": "eightshift-block",
    "maxCols": 12,
    "breakpoints": {
      "mobile": 479,
      "tablet": 1279,
      "desktop": 1919,
      "large": 1920
    },
    "containers": {
      "default": "1330px"
    },
    "gutters": {
      "none": "0",
      "default": "25px",
      "big": "50px"
    },
    "sectionSpacing": {
      "min":  -300,
      "max":  300,
      "step": 10
    },
    "sectionInSpacing": {
      "min":  0,
      "max":  300,
      "step": 10
    },
    "colors": [
      {
        "name": "Infinum",

>         "slug": "infinum",

Reviewer 2: Maybe remove this one?

        "color": "#D8262C"
      },
      {
        "name": "Black",
        "slug": "black",
        "color": "#111111"
      },
      {
        "name": "White",
        "slug": "white",
        "color": "#FFFFFF"
      }
    ]
  }
}
```

**Output:**
```html
<style>
  :root {
    --global-custom-blocks-name: eightshift-block;
    --global-max-cols: 12;
    --global-breakpoints-mobile: 479;
    --global-breakpoints-tablet: 1279;
    --global-breakpoints-desktop: 1919;
    --global-breakpoints-large: 1920;
    --global-containers-default: 1330px;
    --global-gutters-none: 0;
    --global-gutters-default: 25px;
    --global-gutters-big: 50px;
    --global-section-spacing-min: -300;
    --global-section-spacing-max: 300;
    --global-section-spacing-step: 10;
    --global-section-in-spacing-min: 0;
    --global-section-in-spacing-max: 300;
    --global-section-in-spacing-step: 10;

>     --global-colors-infinum: #D8262C;

Reviewer 2: 🆙

    --global-colors-black: #111111;
    --global-colors-white: #FFFFFF;
  }
</style>
```

> To use global variable just call:

Reviewer 2: ```suggestion
You can use a global variable like any other CSS variable:
```

```css

> var(--global-colors-infinum);

Reviewer 2: ```suggestion
color: var(--global-colors-infinum);
```

```

> ## Variable Default

Reviewer 2: ```suggestion
## Variable Defaults
```

> As we said in the beginning, all CSS variables are controlled with the block/component manifest but we didn't want to output all attributes as CSS variables (because you don't need them all), so you need to set what attributes you are going to output.

Reviewer 2: ```suggestion
As we said in the beginning, all CSS variables are defined within the block/component manifest. To avoid outputting all of the variables (because you might not need them all), the attributes that will be output as variables need to be defined.
```

> To output attribute as CSS variable, you need to set the `variable` key to `default`. This way you will get the output of the default value or value set in the database.

Reviewer 2: ```suggestion
To output an attribute as a CSS variable, you need to set the `variable` key to `default`. This way you will get the output of the default value or value set in the database.
```

**Manifest:**
```json
{
  "attributes": {
    "typographySize": {
      "type": "string",
      "default": "120-default",
      "variable": "default"
    }
  },
  "options": {
    "typographySize": [
      {
        "label": "180 Display",
        "value": "180-default"
      },
      {
        "label": "120 Display",
        "value": "120-default"
      },
      {
        "label": "80 Display",
        "value": "80-default"
      }
    ]
  }
}
```

**Output:**
```css
--typography-size: 120-default
```

## Variable Color

> If you have attributes that represent color used from the global variable you can output the CSS variable for color directly.

Reviewer 2: ```suggestion
If you have an attribute that represent colors from global variables, you can output the CSS variable for color directly.
```

**Manifest:**
```json
{
  "attributes": {
    "typographyColor": {
      "type": "string",
      "default": "infinum",
      "variable": "color"
    }
  },
  "options": {
    "typographyColor": [
      {
        "name": "Infinum",
        "slug": "infinum",
        "color": "#D8262C"
      },
      {
        "name": "Black",
        "slug": "black",
        "color": "#111111"
      },
    ]
  }
}
```

**Output:**
```css
--typography-color: var(--global-colors-infinum);
```

> ## Variable Select

Reviewer 2: ```suggestion
## Complex attributes
```

> If you have attributes that are used in the `SelectControl` component by default CSS variable will output the `value` stored in the attribute.

Reviewer 2: ```suggestion
If you have attributes stored as objects (for example used in the `SelectControl` component), by default CSS variable generator will output the `value` stored in the attribute.
```

> If you set the variable to `select`, the CSS variable will look for the `variable` key in the manifest options.

Reviewer 2: ```suggestion
Setting the `variable` key to `select`, the CSS variable generator will look for the `variable` key in the manifest `options` field.
```

> If the `variable` key is not set in options, the fallback will be the `value` key.

Reviewer 2: ```suggestion
If the `variable` key is not set in options, the `value` key will be used as fallback.
```

> > You can write any CSS selector, variable in the custom variable key.

Reviewer 2: ```suggestion
> You can write any CSS declaration or variable in the custom variable key.
```

**Manifest:**
```json
{
  "attributes": {
    "typographySize": {
      "type": "string",
      "default": "120-default",
      "variable": "select"
    }
  },
  "options": {
    "typographySize": [
      {
        "label": "180 Display",
        "value": "180-default",
        "variable": "something"
      },
      {
        "label": "120 Display",
        "value": "120-default",
        "variable": "somethingNew"
      },
      {
        "label": "80 Display",
        "value": "80-default",
        "variable": "somethingNew2"
      }
    ]
  }
}
```

**Output:**
```css
--typography-size: somethingNew
```

> ### Variable Select - Editor only

Reviewer 2: ```suggestion
### Complex variables - Editor only
```

> If you want to show different variables on the frontend and editor you can use the additional key `variableEditor` to output variable only for the editor.

Reviewer 2: ```suggestion
If you want to show different variables on the frontend and in the editor you can use the `variableEditor` key to output a different variable inside the block editor.
```

> This way fronted will get the `variable` key and editor `variableEditor` key.

Reviewer 2: ```suggestion

```

> We have made fallbacks if you don't put `variableEditor` on all the options.

Reviewer 2: ```suggestion
If `variableEditor` is not present on all the options, the fallback behaviour is as follows:
```

> **Fallback tree - editor:**

Reviewer 2: ```suggestion
**Block editor fallback tree:**
```

> * variableEditor

Reviewer 2: ```suggestion
* `variableEditor`
```

> * variable

Reviewer 2: ```suggestion
* `variable`
```

> * value

Reviewer 2: ```suggestion
* `value`
```

> **Fallback tree:**

Reviewer 2: ```suggestion
** Frontend fallback tree:**
```

> * variable

Reviewer 2: ```suggestion
* `variable`
```

> * value

Reviewer 2: ```suggestion
* `value`
```

> ## Variable Boolean

Reviewer 2: ```suggestion
## Boolean variables
```

> If you have attributes that are used in the `ToggleControl` component by default CSS variable will output the `value` stored in the attribute.

Reviewer 2: ```suggestion
By default, the CSS variable generator will output values stored in attributes. That means that boolean variables will get output as variables with `true` and `false` values, which might be something you don't want.
```

> If you set the variable to `boolean`, the CSS variable will look for the `index` key in the manifest options. The first key is the `false` value, and the second is the `true` value.

Reviewer 2: ```suggestion
If you set the `variable` key to `boolean`, the CSS variable generator will look for the `index` array in the manifest options. The first value reprents the `false` value, and the second the `true` value.
```

> If the `variable` key is not set in options fallback does default.

Reviewer 2: ```suggestion
If the `variable` key is not set in options, the fallback is to output the value inside the CSS variable.
```

> > You can write any CSS selector, variable in the custom variable key.

Reviewer 2: ```suggestion
> You can write any CSS declaration or variable in the custom variable key.
```

**Manifest:**
```json
{
  "attributes": {
    "typographyUse": {
      "type": "boolean",
      "default": true,
      "variable": "boolean"
    }
  },
  "options": {
    "typographyUse": [
      "somethingFalse",
      "somethingTrue"
    ]
  }
}
```

**Output:**
```css
--typography-size: somethingTrue
```

> ### Variable Boolean - Editor only

Reviewer 2: ```suggestion
### Boolean variables inside the Block editor
```

> If you want to show different variables on the frontend and editor you can use additional index places. The third key is the `editor false` value and the fourth key is the` editor true` value. Fallback is also in a place like in select.

Reviewer 2: ```suggestion
If you want to show different variables on the frontend and inside the Block editor you can use add values to the `index` key in options. The third value represents the value if false and the fourth editor value if true. Fallback tree is same as in Complex variables.
```

> ## Manual Variables

Reviewer 2: ```suggestion
## Manual variables
```

> We made the option to add custom CSS variables in the output no matter the attributes. To use it, just add top-level key `variables` in the manifest and add each variable as an array item.

Reviewer 2: ```suggestion
There is an option to add custom CSS variables that get output independently from all the attributes. Just add a top-level `variables` key inside the manifest and add each variable as an array item.
```

> Manual variables will just be added at the bottom of the output.

Reviewer 2: ```suggestion
Manual variables will be added at the end of the output.
```

**Manifest:**
```json
{
  "attributes": {
    // ...
  },
  "variables": [
    "--variable1: test1",
    "--variable2: test2",
    "--variable3: test3"
  ]
}
```

**Output:**
```css
--variable1: test1;
--variable2: test2;
--variable3: test3;
```
 

> ## Manual Variables - Editor only

Reviewer 2: ```suggestion
## Manual variables inside the Block editor
```

> If you want to add manual attributes only to the editor you can use the `variablesEditor` key. Everything else applies from the `Manual Variables`.

Reviewer 2: ```suggestion
If you want to add manual variables that only apply inside the Block editor you can use the `variablesEditor` key. Everything works the same as described in the _Manual variables_ section.
```

> If you combine `variablesEditor` and `variables` in one manifest, both of them will be outputted in the editor but only `variables` will be outputted on the frontend.

Reviewer 2: ```suggestion
If you define both `variablesEditor` and `variables`, both will be output in the editor, but only `variables` will be output on the frontend.
```
